Added macro for generating element-wise methods to the SoAs (#47625)
cms-sw/cmssw: DataFormats/SoATemplate/README.md

function computes the first byte of a structure right after a layout, allowing using a single buffer for multiple
layouts.

## Customized methods

> It is possible to generate methods inside the `element` and `const_element` nested structs using the `SOA_METHODS` and `SOA_CONST_METHODS` macros. More than one declaration of these macros are not allowed and all the customized methods can be implemented as macro argument. [An example is showed below.](#examples)

Reviewer: ```suggestion
It is possible to generate methods inside the `element` and `const_element` nested structs using the `SOA_METHODS` and `SOA_CONST_METHODS` macros. Each of these macros can be called only once, and can define multiple methods. [An example is showed below.](#examples)
```

using SoA1LayoutAligned = SoA1LayoutTemplate<cms::soa::CacheLineSize::defaultSize, cms::soa::AlignmentEnforcement::enforced>;
```

It is possible to declare methods that operate on the SoA elements:

```C++
#include "DataFormats/SoALayout.h"

GENERATE_SOA_LAYOUT(SoATemplate,
  SOA_COLUMN(double, x),
  SOA_COLUMN(double, y),
  SOA_COLUMN(double, z),
  
  // methods operating on const_element
  SOA_CONST_METHODS(

>     auto norm() {

Reviewer 2: How about recommending the `const_element` member functions to be `const` along
```suggestion
    auto norm() const {
```
? (also in the tests)

I guess the non-mutability is mostly implied, but to make sure that these functions cane be called from `const const_element`.

      return sqrt(x()*x() + y()+y() + z()*z());
    }
  ),

  // methods operating on element
  SOA_METHODS(
    void scale(float arg) {
      x() *= arg;
      y() *= arg;
      z() *= arg;
    }

>   ),

Reviewer: ```suggestion
  ),

```

>   SOA_SCALAR(int, detectorType)

Reviewer: ```suggestion

  SOA_SCALAR(int, detectorType)
```
